tracer/src/Datadog.Trace/Logging/DirectSubmission/Sink/DatadogSink.cs

```diff
@@ -0,0 +1,165 @@
+                        _logStringBuilder.Capacity = InitialBuilderSizeBytes;
+                    }
+
+                    log.Format(_logStringBuilder, _formatter);
+
+                    // would be nice to avoid this, but can't see a way without direct UTF8 serialization (System.Text.Json)
+                    // Currently a rogue giant message still pays the serialization cost but is then thrown away
+                    var serializedLog = _logStringBuilder.ToString();
+
+                    var logSize = Encoding.UTF8.GetByteCount(serializedLog);
+                    if (logSize > MaxMessageSizeBytes)
+                    {
+                        // Note that the logs intake will actually accept oversized logs, and then truncate them to 1MB
+                        // We could continue to send the log as long as the total size isn't over 5MB, but not sure if
+                        // it's worth it or not, so excluding for now.
+                        _logger.Error<int, string>("Log dropped as too large ({Size} bytes) to send to logs intake: {Log}", logSize, serializedLog);
+                        _oversizeLogCallback?.Invoke(log);
+                        continue;
+                    }
+
+                    var requiredTotalSize = _byteCount + logSize + 1; // + 1 for the separate/suffix
+                    if (requiredTotalSize > _serializedLogs.Length && requiredTotalSize <= MaxTotalSizeBytes)
+                    {
+                        // grow the array
+                        var newSize = Math.Min(_serializedLogs.Length * 2, MaxTotalSizeBytes);
+                        var newArray = new byte[newSize];
+                        Array.Copy(_serializedLogs, 0, newArray, 0, _serializedLogs.Length);
+                        _serializedLogs = newArray;
+                    }
+
+                    if (requiredTotalSize > MaxTotalSizeBytes)
+                    {
+                        // send what we have, add the log to the subsequent chunk
+                        await ReplaceFinalSeparatorAndSendChunk().ConfigureAwait(false);
+                    }
+
+                    // add the log to the batch
+                    var bytesWritten = Encoding.UTF8.GetBytes(serializedLog, 0, serializedLog.Length, _serializedLogs, _byteCount);
+                    Debug.Assert(bytesWritten == logSize, "Actual bytes written should equal the log size");
+
+                    _byteCount += bytesWritten;
+
+                    // add the separator
+                    _serializedLogs[_byteCount] = SeparatorAsUtf8Byte;
+                    _byteCount++;
+                    _logCount++;
+                }
+
+                if (_logCount > 0)
+                {
+                    await ReplaceFinalSeparatorAndSendChunk().ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                _logger.Error(e, "An error occured sending logs to Datadog");
```

Comment:
Do we want to try adding a circuit breaker if this continues failing?

Reply:
Probably a good idea? I'm inclined to leave it to a subsequent PR though, as this one is big enough?

Also, we don't have a circuit breaker for traces... should we?